File: Tests/PUPSISPortalTests/ScheduleStoreTests.swift
import XCTest
@testable import PUPSISPortal

/// Everything here writes to a temp directory — never the real
/// Application Support path, which holds the user's actual schedule.
final class ScheduleStoreTests: XCTestCase {
    private var url: URL!

    override func setUpWithError() throws {
        url = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("ScheduleStoreTests-\(UUID().uuidString)", isDirectory: true)
            .appendingPathComponent("schedule.json")
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: url.deletingLastPathComponent())
    }

    private let fixture = [
        ClassSession(subjectCode: "COMP 20073", description: "Data Structures",
                     faculty: "SANTOS, JUAN", day: .tuesday, start: 14 * 60, end: 16 * 60),
        ClassSession(subjectCode: "COMP 20073", description: "Data Structures",
                     faculty: "SANTOS, JUAN", day: .friday, start: 13 * 60 + 30, end: 16 * 60 + 30),
    ]

    func testRoundTripsSessionsAndTimestamp() {
        let stamp = Date(timeIntervalSince1970: 1_754_400_000)
        ScheduleStore.save(fixture, lastUpdated: stamp, to: url)

        let loaded = ScheduleStore.load(from: url)
        XCTAssertEqual(loaded?.sessions, fixture)
        // Encoded as a Double, so compare with a tolerance rather than ==.
        XCTAssertEqual(loaded?.lastUpdated.timeIntervalSince1970 ?? 0,
                       stamp.timeIntervalSince1970, accuracy: 0.001)
    }

    /// Everything the round-trip drops would silently vanish from the UI.
    func testRoundTripKeepsEveryField() {
        ScheduleStore.save(fixture, to: url)
        let loaded = try? XCTUnwrap(ScheduleStore.load(from: url)?.sessions.first)

        XCTAssertEqual(loaded?.subjectCode, "COMP 20073")
        XCTAssertEqual(loaded?.description, "Data Structures")
        XCTAssertEqual(loaded?.faculty, "SANTOS, JUAN")
        XCTAssertEqual(loaded?.day, .tuesday)
        XCTAssertEqual(loaded?.start, 14 * 60)
        XCTAssertEqual(loaded?.end, 16 * 60)
    }

    func testMissingFileLoadsAsNoCacheRatherThanCrashing() {
        XCTAssertNil(ScheduleStore.load(from: url))
    }

    func testCorruptFileLoadsAsNoCache() throws {
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try Data("not json".utf8).write(to: url)

        XCTAssertNil(ScheduleStore.load(from: url))
    }

    func testDeleteRemovesTheFileFromDisk() {
        ScheduleStore.save(fixture, to: url)
        XCTAssertTrue(FileManager.default.fileExists(atPath: url.path))

        ScheduleStore.delete(at: url)
        XCTAssertFalse(FileManager.default.fileExists(atPath: url.path))
    }

    /// The cache holds the student's own schedule; nobody else on the machine
    /// should be able to read it.
    func testCachedScheduleIsOwnerReadableOnly() throws {
        ScheduleStore.save(fixture, to: url)

        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        XCTAssertEqual(attributes[.posixPermissions] as? Int, 0o600)
    }

    /// Identity is positional — two decodes of the same session must not
    /// produce different ids, or SwiftUI rebuilds every block on refresh.
    func testIdentityIsStableAcrossDecodes() {
        ScheduleStore.save(fixture, to: url)
        let first = ScheduleStore.load(from: url)?.sessions.map(\.id)
        let second = ScheduleStore.load(from: url)?.sessions.map(\.id)

        XCTAssertEqual(first, second)
        XCTAssertEqual(first, fixture.map(\.id))
    }
}

File: windows/PUPSISPortal.Core.Tests/ScheduleStoreTests.cs
using Xunit;
using System;
using System.Collections.Generic;

namespace PUPSISPortal.Core.Tests;

/// <summary>
/// Schedule persistence. Everything here writes to a temp directory — never
/// the real LocalAppData path, which holds the user's actual schedule.
/// </summary>
public class ScheduleStoreTests : IDisposable
{
    private readonly string _testDir;

    public ScheduleStoreTests()
    {
        // Create a unique temp directory for this test
        _testDir = Path.Combine(Path.GetTempPath(), $"ScheduleStoreTests-{Guid.NewGuid()}");
        Directory.CreateDirectory(_testDir);
        // Override the store to use our test directory
        ScheduleStore.SetTestDirectory(_testDir);
    }

    public void Dispose()
    {
        // Reset to default directory
        ScheduleStore.SetTestDirectory(null);
        if (Directory.Exists(_testDir))
            Directory.Delete(_testDir, recursive: true);
    }

    private readonly List<ClassSession> _fixture = new()
    {
        new ClassSession
        {
            SubjectCode = "COMP 20073",
            Description = "Data Structures",
            Faculty = "SANTOS, JUAN",
            Day = Weekday.Tuesday,
            Start = 14 * 60,
            End = 16 * 60
        },
        new ClassSession
        {
            SubjectCode = "COMP 20073",
            Description = "Data Structures",
            Faculty = "SANTOS, JUAN",
            Day = Weekday.Friday,
            Start = 13 * 60 + 30,
            End = 16 * 60 + 30
        }
    };

    [Fact]
    public void RoundTripsSessionsAndTimestamp()
    {
        var stamp = new DateTime(2025, 8, 6, 0, 0, 0, DateTimeKind.Utc);
        ScheduleStore.Save(_fixture, stamp);

        var loaded = ScheduleStore.Load();
        Assert.NotNull(loaded);
        Assert.Equal(_fixture, loaded.Sessions);
        // DateTime comparison with tolerance for serialization precision
        Assert.Equal(stamp.Ticks, loaded.LastUpdated.Ticks, tolerance: 1_000_000); // ~100ms tolerance
    }

    [Fact]
    public void RoundTripKeepsEveryField()
    {
        ScheduleStore.Save(_fixture);
        var loaded = ScheduleStore.Load();

        Assert.NotNull(loaded);
        var session = loaded.Sessions.First();
        Assert.Equal("COMP 20073", session.SubjectCode);
        Assert.Equal("Data Structures", session.Description);
        Assert.Equal("SANTOS, JUAN", session.Faculty);
        Assert.Equal(Weekday.Tuesday, session.Day);
        Assert.Equal(14 * 60, session.Start);
        Assert.Equal(16 * 60, session.End);
    }

    [Fact]
    public void MissingFileLoadsAsNoCacheRatherThanCrashing()
    {
        var loaded = ScheduleStore.Load();
        Assert.Null(loaded);
    }

    [Fact]
    public void CorruptFileLoadsAsNoCache()
    {
        var path = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "PUPSISPortal",
            "schedule.json"
        );
        var dir = Path.GetDirectoryName(path);
        if (dir != null && !Directory.Exists(dir))
            Directory.CreateDirectory(dir);

        File.WriteAllText(path, "not json");
        try
        {
            var loaded = ScheduleStore.Load();
            Assert.Null(loaded);
        }
        finally
        {
            if (File.Exists(path))
                File.Delete(path);
        }
    }

    [Fact]
    public void DeleteRemovesTheFile()
    {
        ScheduleStore.Save(_fixture);
        Assert.NotNull(ScheduleStore.Load());

        ScheduleStore.Delete();
        Assert.Null(ScheduleStore.Load());
    }

    [Fact]
    public void IdentityIsStableAcrossDecodes()
    {
        ScheduleStore.Save(_fixture);
        var first = ScheduleStore.Load()?.Sessions.Select(s => s.Id).ToList();
        var second = ScheduleStore.Load()?.Sessions.Select(s => s.Id).ToList();

        Assert.NotNull(first);
        Assert.NotNull(second);
        Assert.Equal(first, second);
    }
}
